### ISSUE TITLE
Regularization for LTC/CfC; reinitializing hidden state after inference

### ISSUE THREAD

**Issue by off-axis** [mlech26l/ncps#79]
Hi @mlech26l and @raminmh, 

I'm working on a time-series forecast project in Tensorflow/Keras using CfC. I have two questions:

1. I've encountered what I believe is an overfitting problem. Regular actions like reducing learning rate and model complexity didn't help; I started searching for L1/L2 weight regularization parameters in CfC class, but found nothing. I thought about using weight decay, but soon found out that unlike in pytorch, weight decay and L2 regularization are not the same in Keras. Since Keras doesn't provide training loop creation opportunities, it's impossible to calculate L2 manually and directly change the training loss in the loop. So, the question is: do you plan on implementing weight regularization for LNNs, or is there already a way to do this that I missed? Or am I completely wrong, and regularization won't work with LNNs like it does with usual neural networks?

2. From the papers that you have published, I've read that LTC and CfC models are able to shift some of their parameters according to the input data during inference. Does that mean that after each inference batch I would need to save model's weights and then reinitialize the hidden state?

Thank you in advance.

**Comment by MoniFarsang** [comment]
1. For using regularization in LTCs/CfCs, you can extend the `LTCCell` or `CfCCell` classes and add regularizers directly to the weights in their `add_weight()` methods. This way, you can apply L1, L2, or combined L1L2 regularization exactly like in standard keras layers.
For example, in your custom cell you could do something like:
```
self._params["gleak"] = self.add_weight(
    name="gleak",
    shape=(self.state_size,),
    dtype=tf.float32,
    constraint=tf.keras.constraints.NonNeg(),
    initializer=self._get_initializer("gleak"),
    regularizer=tf.keras.regularizers.L1L2(l1=0.01, l2=0.01)  # adjust values as needed
)
```
You can repeat this for each weight you want to regularize. If you only want one type, you can use `tf.keras.regularizers.L1` or `tf.keras.regularizers.L2` instead of `L1L2` or just leave them to their default values, which is 0.0 in L1L2.

2. The model's weights don't change after training, LTC and CfC are fixed-weight networks during inference. The "liquid" aspect refers to the effective time constants in the dynamics, which are computed from some of the learned weights together with the current input and hidden state. Since the input and hidden state vary over time, the computed time constant changes at each step, but the underlying learned parameters remain the same, so you don't have to save them.
Regarding the hidden state, though not sure if this is the case for you: if you're running training/inference on consecutive batches and want to preserve the hidden state between them (instead of resetting it each time), you can set `stateful=True` in the LTC/CfC layer. This behavior is the same as in any other RNN in tensorflow/keras: https://keras.io/api/layers/recurrent_layers/rnn/
